File: backend/app/schemas/review.py
from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field, HttpUrl  # noqa: F401


# --- Request schemas ---


class ReviewCreate(BaseModel):
    pr_url: str = Field(
        ...,
        description="GitHub PR URL or GitLab MR URL",
        examples=[
            "https://github.com/owner/repo/pull/123",
            "https://gitlab.com/group/project/-/merge_requests/456",
        ],
    )
    config: dict = Field(default_factory=dict, description="Optional review configuration overrides")
    template_id: UUID | None = Field(None, description="Optional review template ID; its rules are merged into config")


class WebhookPayload(BaseModel):
    action: str
    number: int
    pull_request: dict
    repository: dict


# --- Response schemas ---


class FindingResponse(BaseModel):
    id: UUID
    agent_role: str
    severity: str
    file_path: str
    line_number: int | None
    title: str
    description: str
    suggested_fix: str | None
    confidence: float
    cost_usd: Decimal
    created_at: datetime

    model_config = {"from_attributes": True}


class ReviewResponse(BaseModel):
    id: UUID
    pr_url: str
    repo_name: str
    pr_number: int | None
    platform: str = "github"
    status: str
    total_issues: int
    total_cost_usd: Decimal
    error_message: str | None
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


class ReviewDetailResponse(ReviewResponse):
    findings: list[FindingResponse] = []
    config: dict = {}


class ReviewListResponse(BaseModel):
    items: list[ReviewResponse]
    total: int


class ReviewEventResponse(BaseModel):
    id: UUID
    event_type: str
    event_data: dict
    timestamp: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/review.ts
export type ReviewStatus = "pending" | "running" | "completed" | "failed" | "cancelled";
export type Severity = "high" | "medium" | "low" | "info";
export type AgentRole = "logic" | "security" | "edge_case" | "convention" | "performance";

export interface Review {
  id: string;
  pr_url: string;
  repo_name: string;
  pr_number: number | null;
  status: ReviewStatus;
  total_issues: number;
  total_cost_usd: number;
  error_message: string | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
}

export interface ReviewDetail extends Review {
  findings: Finding[];
  config: Record<string, unknown>;
}

export interface Finding {
  id: string;
  agent_role: AgentRole;
  severity: Severity;
  file_path: string;
  line_number: number | null;
  title: string;
  description: string;
  suggested_fix: string | null;
  confidence: number;
  cost_usd: number;
  created_at: string;
}

export interface ReviewListResponse {
  items: Review[];
  total: number;
}

// WebSocket event types
export interface WSEvent {
  event: string;
  data: Record<string, unknown>;
}

export interface AgentState {
  role: AgentRole;
  status: "idle" | "running" | "completed" | "error";
  findings_count: number;
  tokens: number;
  cost_usd: number;
}

export interface ReviewEvent {
  id: string;
  event_type: string;
  event_data: Record<string, unknown>;
  timestamp: string;
}
